### DISCUSSION TITLE
Updates on linkastream.co

### DISCUSSION THREAD

**Discussion by keystroke3** [iptv-org/community#237] [Show and tell]
Hello, thank you all for using salvahe.co to find your stream links. This is a quick update on the service and what is planned for the future. 
### Moving to a new domain
After a few weeks in development, I believe the service is mature enough to have its own domain. `Salvahe.co` is my personal domain where I plan on hosting my portfolio and was only a temporary place for testing. The service is now hosted at https://linkastream.co. `Salvahe.co` will continue hosting for 7 days from today until Monday 14th June 23:59 UTC to allow time for maintainers to update the links. Everything will remain the same in terms of usage and only the domain has changed.

### History support
Link A Stream(LaS) now supports history. If you access the service through a browser, you can now access query history on the `history` page. The history is stored on your browser and not the server, so if you use a different browser or device, your history will be lost.

### API
For those who are or want to use LaS links in a script, app or some sort of auto generator, you can. Adding `json=true` or  `&json=true` to the url will return a response in the form of a json object. For now, the only supported has response codes are `200` with `m3ulink` and `400` with an error message. You can see a simple use example in [here](https://github.com/keystroke3/custom-iptv/blob/master/make_playlist.py) I plan to add other request directives and responses later for features discussed below.

### Libera Pay
Donations can be made through Libera Pay now. If you live in Kenya, you can contact me to find out how you can support me through M-Pesa

## Planned Stuff
These are things I am planning but have not implemented yet. They may come soon or may take time to be created, depending on how much support I receive and time I can dedicate to each.

### Dealing with heavy traffic
Two days ago there was a flood of traffic to a point where I had to shut down the server for a few hours. Apologies to anyone who was affected by the outage. To deal with and prevent this from happening again to strategies will be implemented:  
*  A rate limiter.  
Excess traffic from the same origin will be throttled. The request priority will be reduced to prevent spamming. If you notice your requests taking too to complete, be patient and try again after some time.   
* Server side caching  
will also be added so that the server doesn't have to re-check the same stream repeatedly. This is what was happening on that day. There were multiple requests from different origins for the same stream, so caching is the only sensible solution here. The cache will only be of the requests from the last hour because some stream links might be expired by then.   
### Format querying and selection
For people with poor or unstable connection, it is best to have the ability to select stream quality. This is a feature I am planning to add to LaS. It will take some time to implement correctly efficiently so be patient. Format selection will only work with `json` directive set to `true`in query.

### Batch requests
I think the ability to query multiple streams in on request is very useful. I am therefore going to add this feature in the near future. It might also be a way to mitigate traffic overload. I am currently working on a model that is most efficient but due to server limitations, only so many streams that can be resolved at any given time. An update will come soon.

## Conclusion
That's all for now. I will create new discussions or update this with more information. Please give feedback on your usage experience, questions, feedback and more features you want to see. Thank you.

**Comment by keystroke3** [comment]
A quick update. Youtube is now restricting the server from sending requests due to too many requests from the same IP. Just today there have been over 2000 requests made.  Youtube queries will be down for a short while as I try and figure out how to deal with this issue. If you know of one or multiple reliable servers proxy servers, you can share them here so I can use them to bypass the Youtube rate limit.

**Reply by BellezaEmporium** [reply]
> 
> 
> A quick update. Youtube is now restricting the server from sending requests due to too many requests from the same IP. Just today there have been over 2000 requests made. Youtube queries will be down for a short while as I try and figure out how to deal with this issue. If you know of one or multiple reliable servers proxy servers, you can share them here so I can use them to bypass the Youtube rate limit.

Same on my side, maybe use https://proxyscrape.com ? They got an API that we can interrogate every hour for example, and we get fresh proxies everytime.

**Reply by keystroke3** [reply]
Yes, the seem to have a quite a good number of proxies. I will try them out and see if they will work.

**Comment by keystroke3** [comment]
Update:
Caching has now been merged from the staging to production server. Similar requests made in the last hour should now get an instant response. You can test it out and tell me what you find. The server should be able to handle more requests now and have less chance of triggering a 429 error. I can now begin working on format queries and creating adaptive links.

**Comment by keystroke3** [comment]
I hate to be that guy,  but I had to blacklist a client for too many requests. It started with 5 req/hr then to 100 then 500 and now 1278 requests in the last 40 min. Link A Stream is intended for individual use, if you are using my server to serve requests for your website or service, please contact me. I will be happy to create a dedicated instance for you and offer support and extra features if you donate. If you are getting 403 error, then contact me.

**Reply by BellezaEmporium** [reply]
Don't be sorry about it, I had someone that litterally did 1 request / 3 seconds to 1 single channel. That's understandable to take some measures

**Reply by keystroke3** [reply]
I had to restart my server several times because of them. Leading to legit requests and vpn connections served there being dropped. I can tell they are running some sort of website/app/service, because they make requests for the same stream 10 times in a row in 1 min. I am happy to run a separate instance for them, if they fund it.

**Reply by keystroke3** [reply]
The keep changing their IP, so the I will disable the app for a few hours until I can finish writing my script to block all future attempts. I don't want to block innocent request from someone quickly flipping between channels, so I have to make sure the script works properly. Sorry for any inconvenience.

**Reply by keystroke3** [reply]
It's back up. Now if anyone tries spamming, they will automatically get banned. I don't have to do anything. If you believe you have been wrongfully banned, contact me and I will do a manual check.

**Reply by BellezaEmporium** [reply]
> 
> 
> It's back up. Now if anyone tries spamming, they will automatically get banned. I don't have to do anything. If you believe you have been wrongfully banned, contact me and I will do a manual check.

I still see a 403 Forbidden when trying to look after the linkastream root page, and a 404 not found when testing the "headless" endpoint.

**Reply by keystroke3** [reply]
That's strange. I am not seeing any 403 responses in my logs, and I can access the page fine. I am also seeing more traffic from other places. Are you comfortable giving your IP so I can look it up. Maybe it was blocked by the firewall. I have disabled then nginx filters fro now. Try again and tell me what you get.

**Comment by keystroke3** [comment]
I am doing server maintenance and service will be down for about an hour. I apologize for any inconvenience caused.I will update when complete.

**Reply by keystroke3** [reply]
Maintenance complete. Service restored.

**Reply by BellezaEmporium** [reply]
Ah, now works for me.

**Reply by keystroke3** [reply]
Huh! I guess I will never know what the problem was, because I don't know what fixed it. Maybe it was rDNS issues. That's the only network change I made.

**Comment by tarrazuapps** [comment]
This is great stuff.
I am trying to use the API json option but it is not working for me, behaves the same as . The example link is unavailable.

I am trying adding a header "Accept" "application/json" and using this as the get request: https://linkastream.co/search?url=https%3A%2F%2Fwww.makan.org.il%2Flive%2Ftv.aspx%3Fstationid%3D25&json=true

I'm getting HTML in return. What I'm doing wrong?

**Reply by Dum4G** [reply]
>What I'm doing wrong?

Replace search with headless

**Reply by tarrazuapps** [reply]
Thanks!

**Comment by keystroke3** [comment]
I saw some queries about vidio links not being supported. I am working on it now and will update when it is ready.

**Reply by BellezaEmporium** [reply]
https://query-streamlink.lanesh4d0w.repl.co/iptv-query?streaming-ip=http://ruangwaktu.xyz/kampret-loc/local/ruangtv-875.m3u8 ?

(that wouldn't work because those links are expired since some time ago)

**Reply by keystroke3** [reply]
Yes. All m3u links are dynamic. That's why this repo exists in the first place, so they can be refreshed.

**Reply by keystroke3** [reply]
No. The links are usually auto generated by the hosts. They usually have a long UUID or MD5-like string that expires and is hard to guess. Some hosts, like youtube, tie the link to your ip and have an expiry stamp, usually 4-5 hours after generation. Also, each channel has a unique id that you need to know before you even try generating a link. Sometimes, these ids are not publicly viewable before generation. For example, a channel called 'abc' with id '123' on the video platform might have an id of 'xyz123' on the cdn. Each video also has it's own unique id.

**Comment by keystroke3** [comment]
It seems some changes I made to the server caused the firewall to reject all incoming requests. I have fixed it and it works now. I apologize for problems caused by this error.

**Comment by tarrazuapps** [comment]
Just so you guys know. I added an experimental feature in the xbox app to use these services, although I don't fully understand if the web api has been standarized or not. The user has to go to the advanced tab and check an option and provide the streamlink service url.

So the implementation is: if an item in the m3u list has a website defined, it will use the service specified to query the real stream url, then will keep this url cached for 30 minutes or until the full m3u8 playlist is reloaded.

So for example this entry will resolve and play (just the tvg-url part is important)

#EXTINF:-1 channelNumber=1001 tvgId=1001 tvg-id=1001 tvg-url="https://www.youtube.com/watch?v=82wAxR0cTAA" group-title="" tvg-country="VEN" tvgcountry="VEN"tvgname="" tvg-name="Telesur HD", Telesur HD 
https://manifest.googlevideo.com/api/manifest/hls_playlist/expire/1626127445/ei/9WfsYJWIG8qUgAeK-6ko.m3u

**Reply by keystroke3** [reply]
That's great and caching is a good choice. It reduces the load on the servers. If you experience issues with my service, you can raise them here

**Comment by keystroke3** [comment]
I am having some issues with loading my bank card to pay for the server. Expect service interruptions for the next few hours. I don't know how long it will last.

**Reply by keystroke3** [reply]
I was able to fix the bank issue. The server should be up and running now. If you get a 503 error, then wait for about 10 min and try again. If the website can't load at all, then contact me.

**Reply by BellezaEmporium** [reply]
> 
> 
> I was able to fix the bank issue. The server should be up and running now. If you get a 503 error, then wait for about 10 min and try again. If the website can't load at all, then contact me.

High loads on your server it seems, or host's offline. I'll wait a few hours.

**Reply by keystroke3** [reply]
Ok, it is experiencing very high cpu usage. I think the resolver is having a hard time with some queries and sometimes holding on to requests even after resolving. I will have to take the service offline for a while and figure out what is happening. I may have to move to a bigger server. It will take a long time. @LaneSh4d0w , if you know some Nodejs, you could help me figure it out.

**Reply by BellezaEmporium** [reply]
I know a bit of Node.js. The iptv-all server's running using Node :)

**Reply by keystroke3** [reply]
Ok, let me send you an invite

**Reply by keystroke3** [reply]
@LaneSh4d0w , I sent you an invite.

**Reply by BellezaEmporium** [reply]
Recieved and accepted.

**Comment by BellezaEmporium** [comment]
Ever thought about adding a rate limiter ? Sadly it's one of the only
solutions available. When I had one request every 3 seconds, almost felt
the same. Might be also the reason why my server was often down for no
reason.

**Reply by keystroke3** [reply]
I have a rate limiter on nginx. I even ban people who abuse the service, with 122 ips currently blocked. Even with redis caching the server just can't keep up. I will work on the code today and see if I can put back up today.

**Comment by keystroke3** [comment]
For better or for worse, linkastream is up and running now. Right now, very few platforms are supported, but the two most important, youtube and dailymotion are supported. There is no 429 error protection, so my server ip may get banned by the platforms. Now I also have to worry about running out of RAM as well since the new process consumes more RAM the longer the service is up. The best way to ensure the server stays up as long as possible is to donate. It also encourages me to find bugs in the software, fix issues, and find new ways to get streams. If you pledge $20 and above, I will set up a private instance for you on a separate server.

**Reply by keystroke3** [reply]
My server is being blocked by youtube for too many requests. I will turn it back on when the block has been lifted.
